extensions/positron-python/python_files/posit/positron/positron_jedilsp.py

```diff
@@ -113,31 +133,94 @@
 apply_jedi_patches()
 
 
-def _jedi_utils_script(project: Optional[Project], document: TextDocument) -> Script:
+def _jedi_utils_script(project: Optional[Project], document: TextDocument) -> Interpreter:
     """
     Search the caller stack for the server object and return a Jedi Interpreter object.
 
     This lets us use an `Interpreter` (with reference to the shell's user namespace) for all LSP
     methods without having to vendor all of that code from `jedi-language-server`.
     """
-    # Get the server object from the caller's scope.
+    server = _get_server_from_call_stack()
+    if server is None:
+        raise AssertionError("Could not find server object in the caller's scope")
+    return _interpreter(project, document, server.shell)
+
+
+def _get_server_from_call_stack() -> Optional["PositronJediLanguageServer"]:
+    """Search the call stack for the server object."""
+    level = 0
     frame = inspect.currentframe()
-    if frame is not None and frame.f_back is not None:
-        # Get the server object from the caller's scope.
-        server = frame.f_back.f_locals.get("server")
+    while frame is not None and level < 3:
+        server = frame.f_locals.get("server") or frame.f_locals.get("ls")
+        server = getattr(server, "_wrapped", server)
         if isinstance(server, PositronJediLanguageServer):
-            # Return a Jedi Interpreter.
-            return _interpreter(project, document, server.shell)
-    raise AssertionError("Could not find server object in the caller's scope")
+            return server
+        frame = frame.f_back
+        level += 1
+
+    return None
+
+
+@debounce(1, keyed_by="uri")
+def _publish_diagnostics_debounced(
+    server: "PositronJediLanguageServer", uri: str, filename: Optional[str] = None
+) -> None:
+    # Catch and log any exceptions. Exceptions should be handled by pygls, but the debounce
+    # decorator causes the function to run in a separate thread thus a separate stack from pygls'
+    # exception handler.
+    try:
+        _publish_diagnostics(server, uri, filename)
+    except Exception:
+        logger.exception(f"Failed to publish diagnostics for uri {uri}", exc_info=True)
+
+
+# Adapted from jedi_language_server/server.py::_publish_diagnostics.
+def _publish_diagnostics(
+    server: "PositronJediLanguageServer", uri: str, filename: Optional[str] = None
+) -> None:
+    """Helper function to publish diagnostics for a file."""
+    # The debounce decorator delays the execution by 1 second
+    # canceling notifications that happen in that interval.
+    # Since this function is executed after a delay, we need to check
+    # whether the document still exists
+    if uri not in server.workspace.text_documents:
+        return
+    if filename is None:
+        filename = uri
+
+    doc = server.workspace.get_text_document(uri)
 
+    # Comment out magic/shell/help command lines so that they don't appear as syntax errors.
+    source = "\n".join(
```

Comment:
fyi: this wont catch automagic commands, which is turned on by default. I think that's expected, though